### DISCUSSION TITLE
Add class to `mdast-util-to-hast`

### DISCUSSION THREAD

**Discussion by deadcoder0904** [micromark/micromark#40] [Q&A] [answered]
As said [here](https://github.com/mdx-js/mdx/discussions/1325#discussioncomment-141769), I'm trying to add class to `code` tag.

Here's my markup:

```js
const toHast = require('mdast-util-to-hast')
const toHtml = require('hast-util-to-html')
switch (l.type) {
case 'inlineCode': {
	return toHtml(
		toHast(l, {
			handlers: {
				code: (h, node) => {
					let props = {}
					props.className =
						'p-1 overflow-x-auto text-base bg-gray-900 rounded-md text-blue-gray-200'
					return h(node, 'code', props)
				},
			},
		})
	)
	default:
		return toHtml(toHast(l)) // this works without trying to add handler
}
```

This doesn't work. Am I doing it right or is there any other way to add class to `mdast` or `hast`?

I did find [rehype-add-classes](https://github.com/martypdx/rehype-add-classes/) however I'm not getting how to use it in combination with `unified`. I am here currently:

```js
const unified = require('unified')
const toHast = require('mdast-util-to-hast')
const toHtml = require('hast-util-to-html')
const addClasses = require('rehype-add-classes')

switch (l.type) {
	case 'inlineCode': {
		return unified().use(toHast(l)).use(toHtml).use(addClasses, {
			code:
				'p-1 overflow-x-auto text-base bg-gray-900 rounded-md text-blue-gray-200',
		})
	}
	default:
		return toHtml(toHast(l))
}
```

I am not understanding how to do `unified.use` call on `toHast`. How do I do it? I think this might solve the issue :)

**Comment by wooorm** [comment]
This is not the venue to talk about mdast or hast. This is micromark: a low-level project that you are not dealing with.

It might be good to take a step back and read up on what unified is? https://unifiedjs.com/learn/

mdast is markdown. hast is html. Classes are an HTML thing, so I don’t think there’s a need to do anything in mdast space

**Reply by deadcoder0904** [reply]
No, I was gonna open a question on `mdast-util-to-hast` but it redirected me to this repo so I opened it here. I think you need to point `Q&A` to the right repo as it points here now.

I did read the first 3 posts on `Unified`, that's why I have been able to solve the 2 issues I had (styling the `table` tag & the `ul>li` tag)

> mdast is markdown. hast is html.

Yes, I now have it bye-heart.

Here's what I want to do:

Add class to the `code` tag which is inside of a `li` which I've explained in a bit detail [here](https://github.com/mdx-js/mdx/discussions/1352#discussion-38323)

The reason to do this is that right now the `code` tag inside of a `li` is rendered as-is without any styling.

`this gets styled correctly as it uses custom 'Code' tag from 'MDXComponents'`

However, the `code` tag outside of `li` has a particular style as it's passed down via `MDXComponents` which I've also discussed in detail on the above-referrenced issue.

- `this does not get styled because it's inside of a 'li'`

Also, the above `rehype-add-classes` library adds classes to `rehype`.

I just need to know how to use `toHast` in combination with `unified`.

The following works:

```js
toHtml(toHast(l))
```

Can I convert the above code into `unified` API at all? If so, how?

**Reply by wooorm** [reply]
> No, I was gonna open a question on mdast-util-to-hast but it redirected me to this repo so I opened it here. I think you need to point Q&A to the right repo as it points here now.

That repo links to [9 different places](https://github.com/syntax-tree/.github/blob/HEAD/support.md#questions) where you can ask questions, and it depends what you are asking about which one you pick.

---

In short: the code you have in `list` to inject HTML in JSX  isn’t needed and prevents you from doing what you want.
As you are using MDX, well: MDX has a `rehypePlugins` option. You can use `rehype-add-classes` there.

**Reply by wooorm** [reply]
btw `martypdx/rehype-add-classes` hasn’t been updated in 3 years, isn’t used a lot, and the code includes bugs. You can build your own plugin with https://github.com/syntax-tree/hast-util-select and https://github.com/syntax-tree/hast-util-classnames

**Reply by deadcoder0904** [reply]
> That repo links to 9 different places where you can ask questions, and it depends what you are asking about which one you pick.

I'm not really sure about that. See 👇

![image](https://user-images.githubusercontent.com/16436270/100544412-7af25480-327b-11eb-880a-de293ae323e0.png)

> MDX has a rehypePlugins option. You can use rehype-add-classes there.

I did that but it isn't working.

```js
options: {
	remarkPlugins: [withCodeSamples, withSyntaxHighlighting, table, list],
	rehypePlugins: [
		addClasses({
			code: 'p-1 overflow-x-auto text-base bg-gray-900 rounded-md text-blue-gray-200',
		}),
	],
},
```

Used someone else's `rehype-add-classes` which I found through Github:

```js
/**
 * cf. https://github.com/martypdx/rehype-add-classes/blob/master/index.mjs
 *
 * Original plugin causes build issues...
 */
const { selectAll } = require('hast-util-select')

module.exports = (additions) => {
	const adders = Object.entries(additions).map(adder)
	return (node) => adders.forEach((a) => a(node))
}

const adder = ([selector, className]) => {
	const writer = write(className)
	return (node) => selectAll(selector, node).forEach(writer)
}

const write = (className) => ({ properties }) => {
	if (!properties.className) {
		properties.className = [className]
	} else if (Array.isArray(properties.className)) {
		if (properties.className.includes(className)) return
		properties.className.push(className)
	} else {
		properties.className += ` ${className}`
	}
}
```

**Reply by wooorm** [reply]
> I'm not really sure about that. See 👇

Thanks, sorry, I messed up there. That link should go somewhere else, I’ll fix it!

**Reply by wooorm** [reply]
See my comment on `martypdx/rehype-add-classes`: it doesn’t work. (edit: I see you find a slightly different fork, but it’s also buggy)

**Reply by wooorm** [reply]
This is INCORRECT:

```js
rehypePlugins: [
	addClasses({
		code: 'p-1 overflow-x-auto text-base bg-gray-900 rounded-md text-blue-gray-200',
	}),
],
```

To pass options, pass them in an array:

```js
rehypePlugins: [
  [
    addClasses,
    {code: 'p-1 overflow-x-auto text-base bg-gray-900 rounded-md text-blue-gray-200'}
  ]
],
```

**Reply by wooorm** [reply]
Here is a `rehype-classes` plugin that works (should work, I wrote it off the top of my head):

```js
var selectAll = require('hast-util-select').selectAll
var classnames = require('hast-util-classnames')

module.exports = rehypeClasses

function rehypeClasses(options) {
  return transform

  function transform(tree) {
    var selector
    var index

    for (selector in options) {
      nodes = selectAll(selector, tree)
      index = -1
      while (++index < nodes.length) {
        classnames(nodes[index], options[selector])
      }
    }
  }
}
```

**Reply by deadcoder0904** [reply]
And that worked 🎉

Thank you so much @wooorm. The unified ecosystem looks awesome. It is very huge & most of the repos I found were yours so thanks for all the help :)

**Reply by wooorm** [reply]
Awesome, glad to hear!
BTW: its a lot of projects and syntax trees are complex, so it's totally understandable to have questions!

**Comment by deadcoder0904** [comment, marked as answer]
Wow, that new code worked too. I'll post the answer here for quick access:

# rehype/add-classes.js

```js
// Thanks @wooorm -> https://github.com/micromark/micromark/discussions/40#discussioncomment-141871

const selectAll = require('hast-util-select').selectAll
const classnames = require('hast-util-classnames')

function rehypeClasses(options) {
	return transform

	function transform(tree) {
		let selector
		let index

		for (selector in options) {
			nodes = selectAll(selector, tree)
			index = -1
			while (++index < nodes.length) {
				classnames(nodes[index], options[selector])
			}
		}
	}
}

module.exports = rehypeClasses
```

# next.config.js

```js
const mdx = [
	options.defaultLoaders.babel,
	{
		loader: '@mdx-js/loader',
		options: {
			remarkPlugins: [withCodeSamples, withSyntaxHighlighting, table, list],
			rehypePlugins: [
				[
					addClasses,
					{
						code: 'p-1 overflow-x-auto text-base bg-gray-900 rounded-md text-blue-gray-200',
					},
				],
			],
		},
	},
]
```
